src/canal.erl

```diff
@@ -138,6 +144,24 @@ handle_cast(reauth, State = #state{auth = #auth{payload = Payload}}) ->
     end,
     {noreply, State2};
 
+handle_cast(reauth, State = #state{auth = undefined}) ->
+    case ?GET_OPT(credentials) of
+        undefined ->
+            Msg = "canal: not authenticated, and no credentials stored",
+            canal_utils:error_msg(Msg),
+            {stop, {error, no_credentials}, State};
+        Creds ->
+            canal_utils:info_msg("canal: auth using credentials in env"),
+            Payload = make_auth_request(Creds, State),
+            case do_auth(Payload, State) of
+                {ok, Data} ->
+                    {noreply, update_auth(State, Data, Payload)};
+                Err = {error, _} ->
+                    canal_utils:error_msg("canal: auth failed with ~p", [Err]),
+                    {stop, Err, State}
+            end
```

Comment:
This case looks very similar to the other `handle_cast/2` function head.  Can you do a recursive call?